example/stochastic_subgrid/stochastic_subgrid_driver.cpp

```diff
@@ -0,0 +1,204 @@
+//========================================================================================
+// (C) (or copyright) 2020-2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#include <memory>
+#include <string>
+#include <vector>
+
+// Local Includes
+#include "bvals/cc/bvals_cc_in_one.hpp"
+#include "interface/metadata.hpp"
+#include "interface/update.hpp"
+#include "mesh/meshblock_pack.hpp"
+#include "parthenon/driver.hpp"
+#include "refinement/refinement.hpp"
+#include "stochastic_subgrid_driver.hpp"
+#include "stochastic_subgrid_package.hpp"
+
+using namespace parthenon::driver::prelude;
+
+namespace stochastic_subgrid_example {
+
+// *************************************************//
+// define the application driver. in this case,    *//
+// that mostly means defining the MakeTaskList     *//
+// function.                                       *//
+// *************************************************//
+StochasticSubgridDriver::StochasticSubgridDriver(ParameterInput *pin,
+                                                 ApplicationInput *app_in, Mesh *pm)
+    : MultiStageDriver(pin, app_in, pm) {
+  // fail if these are not specified in the input file
+  pin->CheckRequired("parthenon/mesh", "ix1_bc");
+  pin->CheckRequired("parthenon/mesh", "ox1_bc");
+  pin->CheckRequired("parthenon/mesh", "ix2_bc");
+  pin->CheckRequired("parthenon/mesh", "ox2_bc");
+
+  // warn if these fields aren't specified in the input file
+  pin->CheckDesired("parthenon/mesh", "refinement");
+  pin->CheckDesired("parthenon/mesh", "numlevel");
+  pin->CheckDesired("Advection", "cfl");
+  pin->CheckDesired("Advection", "vx");
+  pin->CheckDesired("Advection", "refine_tol");
+  pin->CheckDesired("Advection", "derefine_tol");
+}
+
+using stochastic_subgrid_package::ComputeNumIter;
+
+// See the advection.hpp declaration for a description of how this function gets called.
+TaskCollection StochasticSubgridDriver::MakeTaskCollection(BlockList_t &blocks,
+                                                           const int stage) {
```

Comment:
This method is really long, I would recommend breaking into small local functions.

Reply:
I think the method can also be made const, unless I'm missing something.

Reply:
As per the previous discussion, I prefer to not break up this function into smaller functions because there would be no re-use of those functions. Instead, I added some scoping with header comments to break up the function into manageable sections.

This particular method could be made `const` indeed, however, it's overriding a virtual method in the base class, which is not `const`. I'm not sure if the method in the base class should be const or not. There may be a good reason that we don't want it to be const so that other derived classes have the freedom to do non-`const` things in these methods.

Reply:
There is a workaround, add a const method with the same name to the base class, and have the non constant method in your specefic case call the const method. 

https://riptutorial.com/cplusplus/example/16974/avoiding-duplication-of-code-in-const-and-non-const-getter-methods-

This is non-blocking, but I think it would make it easier to follow in the future, if people are not changing the state.